tensorstore/kvstore/s3/s3_key_value_store_test.cc

```diff
@@ -0,0 +1,265 @@
+#include <vector>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include "absl/flags/flag.h"
+#include "absl/log/absl_log.h"
+#include "absl/status/status.h"
+#include "absl/strings/match.h"
+#include "absl/synchronization/mutex.h"
+#include "absl/synchronization/notification.h"
+#include "absl/time/clock.h"
+#include "absl/time/time.h"
+#include <nlohmann/json.hpp>
+#include "tensorstore/context.h"
+#include "tensorstore/internal/env.h"
+#include "tensorstore/internal/http/curl_handle.h"
+#include "tensorstore/internal/http/curl_transport.h"
+#include "tensorstore/internal/http/http_request.h"
+#include "tensorstore/internal/http/http_response.h"
+#include "tensorstore/internal/json_gtest.h"
+#include "tensorstore/internal/path.h"
+#include "tensorstore/internal/schedule_at.h"
+#include "tensorstore/kvstore/s3/s3_request_builder.h"
+#include "tensorstore/kvstore/generation.h"
+#include "tensorstore/kvstore/kvstore.h"
+#include "tensorstore/kvstore/operations.h"
+#include "tensorstore/internal/subprocess.h"
+#include "tensorstore/kvstore/test_util.h"
+#include "tensorstore/util/execution/execution.h"
+#include "tensorstore/util/execution/sender_testutil.h"
+#include "tensorstore/util/executor.h"
+#include "tensorstore/util/future.h"
+#include "tensorstore/util/status_testutil.h"
+#include "tensorstore/util/str_cat.h"
+
+namespace kvstore = ::tensorstore::kvstore;
+using ::tensorstore::CompletionNotifyingReceiver;
+using ::tensorstore::Context;
+using ::tensorstore::Future;
+using ::tensorstore::KeyRange;
+using ::tensorstore::MatchesJson;
+using ::tensorstore::MatchesStatus;
+using ::tensorstore::StorageGeneration;
+using ::tensorstore::internal::SpawnSubprocess;
+using ::tensorstore::internal::Subprocess;
+using ::tensorstore::internal::SubprocessOptions;
+using ::tensorstore::TimestampedStorageGeneration;
+using ::tensorstore::StrCat;
+using ::tensorstore::internal::ScheduleAt;
+using ::tensorstore::internal_http::GetDefaultHttpTransport;
+using ::tensorstore::internal_http::HttpRequest;
+using ::tensorstore::internal_http::HttpResponse;
+using ::tensorstore::internal_http::HttpTransport;
+using ::tensorstore::internal_http::HttpRequestBuilder;
+using ::tensorstore::internal_http::SetDefaultHttpTransport;
+using ::tensorstore::internal_kvstore_s3::S3Credentials;
+using ::tensorstore::internal_kvstore_s3::S3RequestBuilder;
+using ::tensorstore::internal::GetEnv;
+using ::tensorstore::internal::SetEnv;
+using ::tensorstore::internal::UnsetEnv;
+
+
+ABSL_FLAG(std::string, localstack_binary, "",
+          "Path to the localstack");
+
+namespace {
+
+static constexpr char kAwsAccessKeyId[] = "LSIAQAAAAAAVNCBMPNSG";
+static constexpr char kAwsSecretKeyId[] = "localstackdontcare";
+static constexpr char kBucket[] = "testbucket";
+static constexpr char kAwsRegion[] = "af-south-1";
+static constexpr char kUriScheme[] = "s3";
+static constexpr char kDriver[] = "s3";
+static constexpr char kLocalStackEndpoint[] = "http://localhost:4566";
```

Comment:
The endpoint should be a flag. Otherwise, if localstack were spawned from the test itself, then ideally there would be a --port argument that you could pass, and use tensorstore/internal/http/transport_test_utils.h `PickUnusedPortOrDie()`.